manifests/resource/upstream.pp

```diff
@@ -18,82 +33,121 @@
+#        port   => 3002,
+#      },
+#      'localhost:3003' => {
+#        server => 'localhost',
+#        port   => 3003,
+#      },
+#    },
 #  }
 #
 #  Custom config example to use ip_hash, and 20 keepalive connections
 #  create a hash with any extra custom config you want.
-#  $my_config = {
-#    'ip_hash'   => '',
-#    'keepalive' => '20',
-#  }
 #  nginx::resource::upstream { 'proxypass':
-#    ensure              => present,
-#    members => [
-#      'localhost:3000',
-#      'localhost:3001',
-#      'localhost:3002',
-#    ],
-#    upstream_cfg_prepend => $my_config,
+#    ensure    => present,
+#    members   => {
+#      'localhost:3001' => {
+#        server => 'localhost',
+#        port   => 3000,
+#      },
+#      'localhost:3002' => {
+#        server => 'localhost',
+#        port   => 3002,
+#      },
+#      'localhost:3003' => {
+#        server => 'localhost',
+#        port   => 3003,
+#      },
+#    },
+#    ip_hash   => true,
+#    keepalive => 20,
 #  }
+#
 define nginx::resource::upstream (
-  Optional[Array] $members                  = undef,
-  $members_tag                              = undef,
-  Enum['present', 'absent'] $ensure         = 'present',
-  Optional[Hash] $upstream_cfg_append       = undef,
-  Optional[Hash] $upstream_cfg_prepend      = undef,
-  $upstream_fail_timeout                    = '10s',
-  $upstream_max_fails                       = undef,
-  Enum['http', 'stream'] $upstream_context  = 'http',
+  Enum['present', 'absent']           $ensure                 = 'present',
+  Enum['http', 'stream']              $context                = 'http',
+  Nginx::UpstreamMembers              $members                = {},
+  Optional[String]                    $members_tag            = undef,
+  Nginx::UpstreamMemberDefaults       $member_defaults        = {},
+  Optional[String]                    $hash                   = undef,
+  Boolean                             $ip_hash                = false,
+  Optional[Integer[1]]                $keepalive              = undef,
+  Optional[Integer[1]]                $keepalive_requests     = undef,
+  Optional[Nginx::Time]               $keepalive_timeout      = undef,
+  Boolean                             $least_conn             = false,
+  Optional[Nginx::UpstreamLeastTime]  $least_time             = undef,
+  Boolean                             $ntlm                   = false,
+  Optional[Integer]                   $queue_max              = undef,
+  Optional[Nginx::Time]               $queue_timeout          = undef,
+  Optional[String]                    $random                 = undef,
+  Optional[Stdlib::Unixpath]          $statefile              = undef,
+  Optional[Nginx::UpstreamSticky]     $sticky                 = undef,
+  Optional[Nginx::UpstreamZone]       $zone                   = undef,
+  Hash                                $cfg_append             = {},
+  Hash                                $cfg_prepend            = {},
```

Comment:
Can you enforce the datatypes within the hash please?

Reply:
I've added a datatype for this.